File: solve/cpp/heap.cc
#include "common.hh"

void recurse(int k)
{
	if (k == 0)
	{
		output();
	} 
    else
    { 
        if (k&1) 
        {
            for (int i = 0; i != k; ++i) 
            {
                recurse(k - 1);
                swap(a[i], a[k]);
            } 
        }
        else
        {
            for (int i = 0; i != k; ++i) 
            {
                recurse(k - 1);
                swap(a[0], a[k]);
            } 
        } 
        recurse(k - 1); 
    }
}

int main(int argc, char *argv[])
{
	if (init(argc, argv, UNORDERED) == 0)
    { 
        recurse(n - 1);
        return final();
    } 
    return 1;
}

File: solve/lua/heap.lua
#! /usr/bin/lua

require "common"

function recurse(k)
    if k == 1 then
        output()
    else
        if k % 2 == 0 then
            for i = 1, k - 1 do
                recurse(k - 1)
                swap(i, k)
            end
        else
            for i = 1, k - 1 do
                recurse(k - 1)
                swap(1, k) 
            end
        end
        recurse(k - 1)
    end
end

skeleton(UNORDERED, function ()
    recurse(n) 
end)
